### ISSUE TITLE
Upgrade from v1.15.X git to v1.16.x 500 error

### ISSUE THREAD

**Issue by TechCare-Aus** [osTicket/osTicket#6061]
### Prerequisites

* [X] Can you reproduce the problem in a fresh installation of the "develop" branch?
* [X] Do you have any errors in the PHP error log, or javascript console?
* [X] Did you check the [osTicket forums](https://forum.osticket.com/)?
* [X] Did you [perform a cursory search](https://github.com/osTicket/osTicket/issues) to see if your bug or enhancement is already reported?

For more information on how to write a good [bug report](https://github.com/osTicket/osTicket/wiki/Github-Issues-Guidelines)

### Description

Upgrade from Git Develop as of Monday 7th Feb 2022 to the 1.16.x brings osTicket down with error 500.

Apache Log:

```
[root@server logs]# tail -f help_USERNAME_net_au.php.error.log
  thrown in /home/USERNAME/public_html/help/bootstrap.php on line 356
[09-Feb-2022 03:38:43 UTC] PHP Warning:  require(/class.osticket.php): Failed to open stream: No such file or directory in /home/USERNAME/public_html/help/bootstrap.php on line 356
[09-Feb-2022 03:38:43 UTC] PHP Fatal error:  Uncaught Error: Failed opening required '/class.osticket.php' (include_path='./:/:/pear/') in /home/USERNAME/public_html/help/bootstrap.php:356
Stack trace:
#0 /home/USERNAME/public_html/help/main.inc.php(23): require()
#1 /home/USERNAME/public_html/help/scp/staff.inc.php(20): require_once('/home/USERNAME/...')
#2 /home/USERNAME/public_html/help/scp/tickets.php(17): require('/home/USERNAME/...')
#3 /home/USERNAME/public_html/help/scp/index.php(17): require('/home/USERNAME/...')
#4 {main}
  thrown in /home/USERNAME/public_html/help/bootstrap.php on line 356
```

### Steps to Reproduce

1. git checkout 1.16.x
2. git fetch
3. git merge
4. php manage.php deploy /home/USERNAME/public_html/help/ -v

**This is a completely fresh install.**

1. Setup a new git clone to a sub folder on my existing installation (normally https://help.business.com - so https://help.business.com/test
2. php manage.php deploy /home/techcare/public_html/help/test/ -v
3. setup new database and user ready for install
4. navigate to https://help.business.com/test/scp
5. error display "**Error loading settings. Contact admin.**"

**Expected behavior:** [What you expected to happen]

osTicket would, on navigating to site, begin the upgrade process.

**Actual behavior:** [What actually happened]

osTicket shows error 500

### Versions

Was running the latest git prior to most recent commits. So probably 1.15.something.

Server is running latest WHM/cPanel - CentOS v7.9
PHP v8.0 and also tried v8.1 (almost all PHP options on)

### Thoughts

- What are the requirements for osTicket v1.16?? As in what PHP version works or is known to work?
- What other logs would be helpful?
- Can osTicket provide more logging?

**Comment by JediKev** [comment]
@TechCare-Aus 

It appears your `INCLUDE_DIR`constant is not defined for some reason. It gets [defined here](https://github.com/osTicket/osTicket/blob/develop/bootstrap.php#L328) in `bootstrap.php` and then it gets added to `include_path` in [the PHP INI](https://github.com/osTicket/osTicket/blob/develop/bootstrap.php#L354). 

Maybe you’re having some weird PHP issue or perhaps permissions issues (just guessing)? I am unable to replicate this using Apache, PHP 8.0, 1.16.x branch, and cli deploy command (literally just tested to make sure).

**P.S.**
v1.16 requires PHP 8.0. I haven’t tested with the new PHP 8.1 yet but I assume it will be compatible.

Cheers.

**Comment by TechCare-Aus** [comment]
@JediKev - Thanks for the quick response mate.

Yeah I was thinking something about the INCLUDE_DIR was messed up. But how to go about figuring it out...
But why work fine with v1.15.4-20-gfc99cfc2 (fc99cfc)??

![image](https://user-images.githubusercontent.com/2740134/153128027-fb1c203b-ad21-4ba6-bdb7-3f9d22eefec6.png)

I have tried both PHP 8.0 and 8.1 just to see. No difference.

I've reverted to backup. Then tried updating like a usually do, but changing to the 1.15.x branch:

1. git checkout 1.15.x
2. git fetch
3. git merge
4. php manage.php deploy /home/USERNAME/public_html/help/ -v

Now I got the error:

```
[USERNAME@SERVER osTicket]$ php manage.php deploy /home/USERNAME/public_html/help/ -v
Fatal error: Uncaught Error: Non-static method Misc::currentURL() cannot be called statically in /home/USERNAME/osTicket_1.15.5/osTicket/bootstrap.php:369
Stack trace:
#0 /home/USERNAME/osTicket_1.15.5/osTicket/manage.php(24): require_once()
#1 {main}
  thrown in /home/USERNAME/osTicket_1.15.5/osTicket/bootstrap.php on line 369
```

**Comment by ntozier** [comment]
Seems that you neglected to upgrade to PHP 8.  osTicket version 1.16 requires PHP 8.

**Comment by TechCare-Aus** [comment]
@ntozier - I wish that were it... simple fix then :)

![image](https://user-images.githubusercontent.com/2740134/153209671-f8a965de-8bef-4a67-a9f9-cc103cffcdff.png)

**Comment by TechCare-Aus** [comment]
@ntozier - I can see why you would think I didn't however with the screen shot above that. But that was a screen shot BEFORE i upgraded to PHP v8.0 ...

**Comment by ntozier** [comment]
Hrm.  I missed that line.

If it were me I would try downloading the archive from osticket.com/download and see if you have the same problem?

**Comment by TechCare-Aus** [comment]
@ntozier @JediKev - I tried the advice given by ntozier and I got further. At least I was able to log in and run the database upgrade. Here is screen shot of the my installation.

![image](https://user-images.githubusercontent.com/2740134/153529977-819c15eb-cad7-4efe-ac72-de017c8b17e8.png)

The next issue is that I'm not seeing many pages all with the same javascript file = ... **jquery-3.5.1.min.js**

![image](https://user-images.githubusercontent.com/2740134/153530189-a96558bf-8a95-4903-b56f-efd5f39088f1.png)

**PHP ERROR LOG FILE**

```
[11-Feb-2022 02:48:52 UTC] PHP Fatal error:  Uncaught ValueError: Path cannot be empty in /home/USERNAME/public_html/develop/include/class.translation.php:568
Stack trace:
#0 /home/USERNAME/public_html/develop/include/class.translation.php(352): Translation->load_tables()
#1 /home/USERNAME/public_html/develop/include/class.translation.php(380): gettext_reader->get_plural_forms()
#2 /home/USERNAME/public_html/develop/include/class.translation.php(414): gettext_reader->select_string(3)
#3 /home/USERNAME/public_html/develop/include/class.i18n.php(576): gettext_reader->ngettext('template', 'templates', 3)
#4 /home/USERNAME/public_html/develop/include/staff/templates.inc.php(40): _N('template', 'templates', 3)
#5 /home/USERNAME/public_html/develop/scp/templates.php(188): require('/home/USERNAME/...')
#6 {main}
  thrown in /home/USERNAME/public_html/develop/include/class.translation.php on line 568
```

???

**Comment by JediKev** [comment]
@TechCare-Aus 

Try applying the below changes to see if they resolve your issue:
- https://github.com/osTicket/osTicket/pull/6065
- https://github.com/osTicket/osTicket/pull/6068

These will be included in the next maintenance release.

Cheers.

**Comment by TechCare-Aus** [comment]
@JediKev 

I've tried all those commits, but still having similar issues.

If I navigate to /scp/filters.php it shows a 500 error (no further info on the web development console).

PHP error shows:

```
[14-Feb-2022 01:36:14 UTC] PHP Fatal error:  Uncaught ValueError: Path cannot be empty in /home/USERNAME/public_html/develop/include/class.translation.php:568
Stack trace:
#0 /home/USERNAME/public_html/develop/include/class.translation.php(352): Translation->load_tables()
#1 /home/USERNAME/public_html/develop/include/class.translation.php(380): gettext_reader->get_plural_forms()
#2 /home/USERNAME/public_html/develop/include/class.translation.php(414): gettext_reader->select_string(2)
#3 /home/USERNAME/public_html/develop/include/class.i18n.php(576): gettext_reader->ngettext('selected filter', 'selected filter...', 2)
#4 /home/USERNAME/public_html/develop/include/staff/filters.inc.php(174): _N('selected filter', 'selected filter...', 2)
#5 /home/USERNAME/public_html/develop/scp/filters.php(160): require('/home/USERNAME/...')
#6 {main}
  thrown in /home/USERNAME/public_html/develop/include/class.translation.php on line 568
```

**Comment by TechCare-Aus** [comment]
@JediKev - would it help to have access to the development site I'm testing this on?

**Comment by JediKev** [comment]
@TechCare-Aus 

To me it seems like permissions issues. The statement is complaining about `(include $this->STREAM);` and STREAM is the PHAR file (or built-in en_US files). So if it can’t include the PHAR file or built-in files it seems like permission issues or something similar. 

Cheers.

**Comment by TechCare-Aus** [comment]
@JediKev 

Like a permissions problem in the plugins directory?
Owner is same as USERNAME.

![image](https://user-images.githubusercontent.com/2740134/153791278-383193b2-76b3-46ca-84dc-bb100b73d8d1.png)

**Comment by JediKev** [comment]
@TechCare-Aus 

No, languages folder. Do you have any language packs installed or just default English?

Cheers.

**Comment by TechCare-Aus** [comment]
![image](https://user-images.githubusercontent.com/2740134/153791561-a2ed83f4-aeb6-425c-9b4a-7f32085f7b7b.png)

**Comment by TechCare-Aus** [comment]
Weird thing is I've used a completely fresh copy of osTicket v1.16 downloaded direct from https://osticket.com then uploaded and it should have all the right permissions.

Can I turn on more verbose logging?

**Comment by JediKev** [comment]
@TechCare-Aus 

That’s so strange. I have no idea what’s going on then as I can’t replicate what you’re seeing. Turning on verbose logging could help, wouldn’t hurt to try.

Other things you can try is checking your server logs to see if something is being blocked by your host’s security rules, seeing if you have multiple PHP versions/extensions competing with each other, etc.

Cheers.

**Comment by TechCare-Aus** [comment]
@JediKev 

I went through the server logs and found nothing. Only what I've sent you.

What you said about the language however gave me a thought of changing the language to default = en_US.

That removed the 500 errors.

I had missing plugins and the en_GB doesn't show up anymore. Even if I place new .phar plugins into the directory... does the name of the plugin get added some place in the database so that it is not an option to install again?

Are there v1.16 plugins available?

I mentioned it here - https://forum.osticket.com/d/100543-unable-to-upgrade-to-116x-php-issue

**Comment by JediKev** [comment]
@TechCare-Aus 

The plugins currently on our website are compatible with all osTicket versions. If you don’t see them for install I assume you have some kind of permissions issue or your PHP is having issues reading the PHAR files.

I noticed you said you are missing en_GB? Were you using en_GB or en_US when experiencing the 500 errors earlier?

Cheers.

**Comment by TechCare-Aus** [comment]
@JediKev

> I noticed you said you are missing en_GB? Were you using en_GB or en_US when experiencing the 500 errors earlier?

System was set to use en_GB. I changed it to en_US via phpmyadmin. Might have been able to change it in settings however, didn't check first :)

I figured out why some .phar weren't showing up. The Wiki doesn't mention WHERE to put the language.phar file... I presumed it would go in the plugins. But it needs to go in the i18n directory which is mentioned on the download page at the top BEFORE you download.

That explains why I couldn't see anything in the install.

As for the rest it seems like it is all work again.

I changed the locale back to en_AU (probably made zero difference to the original problem).

So I'm thinking that the whole problem began with a missing en_GB.phar in the i18n directory.

This is my development site:

![image](https://user-images.githubusercontent.com/2740134/153797587-3e6f176d-5430-4668-a649-06cc7beffb35.png)

This is my live site:

![image](https://user-images.githubusercontent.com/2740134/153797683-d204d07e-a5c4-471f-9505-a8e5b20337d3.png)

The reason for all the languages is because I was typically running the latest git from develop.

**Comment by TechCare-Aus** [comment]
@JediKev 

Thought I was out of the woods but still seeing random PHP errors (this was just using the search function):

`[14-Feb-2022 08:00:34 UTC] PHP Fatal error:  Uncaught Error: preg_match(): Argument #3 ($matches) cannot be passed by reference in /home/USERNAME/public_html/help/include/                                ajax.tickets.php:58
Stack trace:
#0 /home/USERNAME/public_html/help/include/class.dispatcher.php(145): TicketsAjaxAPI->lookup()
#1 /home/USERNAME/public_html/help/include/class.dispatcher.php(38): UrlMatcher->dispatch('lookup', Array)
#2 /home/USERNAME/public_html/help/include/class.dispatcher.php(120): Dispatcher->resolve('lookup', Array)
#3 /home/USERNAME/public_html/help/include/class.dispatcher.php(38): UrlMatcher->dispatch('/tickets/lookup', NULL)
#4 /home/USERNAME/public_html/help/scp/ajax.php(316): Dispatcher->resolve('/tickets/lookup')
#5 {main}
  thrown in /home/USERNAME/public_html/help/include/ajax.tickets.php on line 58`

If you want me to make a new issue I can?

**Comment by Gerasimos-MDIT** [comment]
Hi all,
I was facing the same issues and I managed to fix it by deleting the plugins. I hope this helps someone.
